File: musicx-domain/src/main/java/fr/xahla/musicx/domain/service/apiHandler/ExternalApiHandler.java
package fr.xahla.musicx.domain.service.apiHandler;

import org.json.JSONObject;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

import static fr.xahla.musicx.domain.application.AbstractContext.logger;

/**
 * Common methods for external API handler classes
 * @author Cochetooo
 * @since 0.3.0
 */
public abstract class ExternalApiHandler {

    protected String apiUrl;

    protected ExternalApiHandler(final String apiUrl) {
        this.apiUrl = apiUrl;
    }

    /**
     * @since 0.3.0
     */
    protected abstract String makeURL(
        final String method,
        final Map<String, String> parameters
    );

    /**
     * @return A JSONObject with the content of the response, otherwise an empty JSONObject.
     * @since 0.3.0
     */
    protected JSONObject sendRequest(final String requestUrl) {
        try (final var httpClient = HttpClient.newHttpClient()) {
            final var request = HttpRequest.newBuilder()
                .uri(URI.create(requestUrl))
                .build();

            final var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            return new JSONObject(response.body());
        } catch (final IOException | InterruptedException exception) {
            logger().error(exception, "API_CALL_ERROR", requestUrl);

            return new JSONObject();
        }
    }

}

File: Musicx/Services/ExternalApi/ExternalApiHandler.cs
using System.Net.Http;
using log4net;
using Musicx.Views.AudioPlayer;
using Newtonsoft.Json.Linq;

namespace Musicx.Services.ExternalApi;

public abstract class ExternalApiHandler
{
    private static readonly ILog Logger = LogManager.GetLogger(typeof(ExternalApiHandler));
    
    protected string ApiUrl;

    protected ExternalApiHandler(string apiUrl)
    {
        ApiUrl = apiUrl;
    }

    protected abstract string MakeUrl(string method, Dictionary<string, string> parameters);

    protected async Task<JObject> SendRequestAsync(string requestUrl)
    {
        using var httpClient = new HttpClient();
        
        try
        {
            var response = await httpClient.GetAsync(requestUrl);

            if (response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                return JObject.Parse(content);
            }
                   
            Logger.Error($"Api Call Error: {requestUrl}");
            return new JObject();
        }
        catch (HttpRequestException httpRequestException)
        {
            Logger.Error($"Request Error: {httpRequestException.Message}");
            return new JObject();
        }
        catch (Exception ex)
        {
            Logger.Error($"Unknown Error: {ex.Message}");
            return new JObject();
        }
    }
}